src/snowflake/snowpark/_internal/compiler/query_generator.py

```diff
@@ -116,18 +148,65 @@ def do_resolve_with_resolved_children(
                     for x in logical_plan.clustering_exprs
                 ],
                 logical_plan.comment,
-                resolved_child,
+                copied_resolved_child,
                 logical_plan,
                 self.session._use_scoped_temp_objects,
                 logical_plan.is_generated,
                 self._snowflake_create_table_plan_info.child_attributes,
             )
 
-        if isinstance(logical_plan, Selectable):
+        elif isinstance(
+            logical_plan,
+            (
+                CreateViewCommand,
+                TableUpdate,
+                TableDelete,
+                TableMerge,
+                CopyIntoLocationNode,
+            ),
+        ):
+            from snowflake.snowpark._internal.compiler.utils import (
+                get_snowflake_plan_queries,
+            )
+
+            # for CreateViewCommand, TableUpdate, TableDelete, TableMerge and CopyIntoLocationNode,
+            # the with definition must be generated before create, update, delete, merge and copy into
+            # query.
+            resolved_child = resolved_children[logical_plan.children[0]]
+            copied_resolved_child = copy.deepcopy(resolved_child)
+            final_queries = get_snowflake_plan_queries(
+                copied_resolved_child, self.resolved_with_query_block
+            )
+            copied_resolved_child.queries = final_queries[PlanQueryType.QUERIES]
+            resolved_children[logical_plan.children[0]] = copied_resolved_child
+            resolved_plan = super().do_resolve_with_resolved_children(
+                logical_plan, resolved_children, df_aliased_col_name_to_real_col_name
+            )
+
+        elif isinstance(logical_plan, Selectable):
             # overwrite the Selectable resolving to make sure we are triggering
             # any schema query build
-            return logical_plan.get_snowflake_plan(skip_schema_query=True)
+            resolved_plan = logical_plan.get_snowflake_plan(skip_schema_query=True)
 
-        return super().do_resolve_with_resolved_children(
-            logical_plan, resolved_children, df_aliased_col_name_to_real_col_name
-        )
+        elif isinstance(logical_plan, WithQueryBlock):
+            resolved_child = resolved_children[logical_plan.children[0]]
+            # record the CTE definition of the current block
+            if logical_plan.name not in self.resolved_with_query_block:
+                self.resolved_with_query_block[
+                    logical_plan.name
+                ] = resolved_child.queries[-1].sql
+
```

Comment:
I'm curious to see if this works well for nested CTE - case where one block contains another block. do we have any test for that?

Reply:
yes, @sfc-gh-jdu have a test case for one cte depends on the other cte here https://github.com/snowflakedb/snowpark-python/blob/main/tests/integ/test_cte.py#L139. However, we do not generate nested CTE statement, they are all flattened with proper order, i believe that was due to a performance concern.

Reply:
yes